### PR TITLE
WIP Incorporate Term Frequency Excerpts

### PR DESCRIPTION
**Pull request by lbschanno**
This pull request is for the current work of incorporating Term Frequency Excerpts (TFEs) into query results, and is currently still a WIP. One of the primary requirements for being able to retrieve TFEs is to identify and track the start and end offsets for each phrase that is a match for the target fields.

I have modified the termOffsetMap var to be a TermOffsetMap object that will both contain the term frequency lists and track phrase offsets. Currently phrase offsets are being tracked for all content:phrase functions and content:within functions, and I have updated tests to verify the results.

I am primarily asking reviewers to take a second glance at the test assertions to see if there are any phrase offsets oddities that do not look correct, or any edge cases that need to be tested against.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java [hunk 1]
```diff
@@ -85,17 +85,17 @@ public Document document() {
         Map<String,Object> map = new HashMap<>();
         Map<String,TermFrequencyList> termOffsetMap = Maps.newHashMap();
         
-        Map<String,TermFrequencyList> termOffsetMap1 = (Map<String,TermFrequencyList>) (map1.get(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME));
-        Map<String,TermFrequencyList> termOffsetMap2 = (Map<String,TermFrequencyList>) (map2.get(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME));
+        TermOffsetMap termOffsetMap1 = (TermOffsetMap) map1.get(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME);
+        TermOffsetMap termOffsetMap2 = (TermOffsetMap) (map2.get(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME));
         
         if (termOffsetMap1 == null) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java:91]: Maybe this routine could be achieved by moving the termfrequency merge into the TermOffsetClass as a mergeTermFrequencies(TermOffsetMap) method or something like that.

Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java:91]: Should be straightforward to do, I'll make the change.

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/ContentOrderedEvaluator.java [hunk 2]
```diff
@@ -243,6 +249,11 @@ private boolean isConverged(List<NavigableSet<EvaluateTermPosition>> offsets, in
                 second = offsets.get(secondIndex);
             } else {
                 // nope
+                
+                // Establish the end offset of the phrase.
+                endOffset = second.first().termWeightPosition.getOffset();
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/ContentOrderedEvaluator.java:254]: This looks correct to me.  +1

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/ContentUnorderedEvaluator.java [hunk 3]
```diff
@@ -215,6 +219,8 @@ public boolean findMatch() {
                 OffsetList o = offsetQueue.remove();
                 
                 if (maxOffset.get().getLowOffset() - o.getMinOffset().getOffset() <= distance) {
+                    // Track the start and end offset for the phrase.
+                    termOffsetMap.putPhraseOffset(field, o.getMinOffset().getOffset(), maxOffset.get().getLowOffset());
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/ContentUnorderedEvaluator.java:223]: Good. +1

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 4]
```diff
@@ -190,14 +204,15 @@ public void reverseSharedTokenIndex() {
...
-        termOffSetMap.put("c", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t3)));
+        termOffSetMap.putTermFrequencyList("a", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t1)));
+        termOffSetMap.putTermFrequencyList("b", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t2)));
+        termOffSetMap.putTermFrequencyList("c", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t3)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, true));
+        assertPhraseOffset("CONTENT", 252, 253);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:215]: This looks correct to me.

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 5]
```diff
@@ -171,13 +184,14 @@ public void testEvaluation1() {
...
-        termOffSetMap.put("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
-        termOffSetMap.put("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
+        termOffSetMap.putTermFrequencyList("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
+        termOffSetMap.putTermFrequencyList("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, true));
+        assertPhraseOffset("CONTENT", 3, 4);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:194]: I don't get this one.... hmm

Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:194]: Oh, I get it now.  The 2 around the 6 means that that token covers the offsets 4 through 8.  Hence 3, 4 being a match.

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 6]
```diff
@@ -210,14 +225,15 @@ public void forwardSharedTokenIndex() {
...
-        termOffSetMap.put("c", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t3)));
+        termOffSetMap.putTermFrequencyList("a", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t1)));
+        termOffSetMap.putTermFrequencyList("b", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t2)));
+        termOffSetMap.putTermFrequencyList("c", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t3)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, true));
+        assertPhraseOffset("CONTENT", 251, 252);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:236]: good

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 7]
```diff
@@ -230,14 +246,15 @@ public void reverseAllSharedTokenIndex() {
...
-        termOffSetMap.put("c", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t3)));
+        termOffSetMap.putTermFrequencyList("a", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t1)));
+        termOffSetMap.putTermFrequencyList("b", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t2)));
+        termOffSetMap.putTermFrequencyList("c", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), t3)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, true));
+        assertPhraseOffset("CONTENT", 252, 252);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:257]: good

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 8]
```diff
@@ -250,14 +267,15 @@ public void forwardAllSharedTokenIndex() {
...
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:278]: good

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 9]
```diff
@@ -272,13 +290,14 @@ public void testEvaluationNoContentFields() {
...
-        termOffSetMap.put("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", false, eventId), list1)));
-        termOffSetMap.put("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", false, eventId), list2)));
+        termOffSetMap.putTermFrequencyList("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", false, eventId), list1)));
+        termOffSetMap.putTermFrequencyList("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", false, eventId), list2)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, false));
+        assertPhraseOffsetsEmpty();
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:300]: good

Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:300]: the within function is always tricky as it always seems to be 1 off.  3,4 is actually a distance of 2 in this regard.

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 10]
```diff
@@ -290,13 +309,14 @@ public void testQuotedEvaluation() {
...
-        termOffSetMap.put("dog's", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
-        termOffSetMap.put("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
+        termOffSetMap.putTermFrequencyList("dog's", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
+        termOffSetMap.putTermFrequencyList("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, true));
+        assertPhraseOffset("CONTENT", 2, 3);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:319]: good

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 11]
```diff
@@ -319,13 +339,14 @@ public void testEvaluation1_1() {
...
-        termOffSetMap.put("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
-        termOffSetMap.put("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
+        termOffSetMap.putTermFrequencyList("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
+        termOffSetMap.putTermFrequencyList("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, true));
+        assertPhraseOffset("CONTENT", 1, 2);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:349]: good.

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java (hunk 9 above)
Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:300]: I am guessing that the within distance has to include the entire range of 7-3 to 7+3.

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java [hunk 12]
```diff
@@ -337,13 +358,14 @@ public void testEvaluation2() {
...
-        termOffSetMap.put("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
-        termOffSetMap.put("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
+        termOffSetMap.putTermFrequencyList("dog", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list1)));
+        termOffSetMap.putTermFrequencyList("cat", new TermFrequencyList(Maps.immutableEntry(new Zone("CONTENT", true, eventId), list2)));
         
         context.set(Constants.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME, termOffSetMap);
         Object o = expr.evaluate(context);
         
         Assert.assertTrue(expect(o, false));
+        assertPhraseOffsetsEmpty();
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:368]: good.  OK, I am going to stop checking each one now.  I think I am convinced that you are pulling the offsets from the correct places.

Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:368]: If there is one that you are particularly concerned about, point it out for me please.

Comment (lbschanno) [reply warehouse/query-core/src/test/java/datawave/query/jexl/functions/ContentFunctionsTest.java:368]: Excellent, will do.
